File: backend/ride-go/internal/application/usecase/accept_ride.go
package usecase

import (
	"errors"

	"github.com/google/uuid"
	"github.com/leonardograselalmeida/fake_uber/internal/application/repository"
)

type AcceptRide struct {
	AccountRepository repository.AccountRepositoryInterface
	RideRepository    repository.RideRepositoryInterface
}

type AcceptRideInput struct {
	DriverId uuid.UUID
	RideId   uuid.UUID
}

func (a *AcceptRide) Execute(input AcceptRideInput) error {
	account, accountError := a.AccountRepository.GetAccountById(input.DriverId)

	if accountError != nil {
		return accountError
	}

	if account == nil {
		return errors.New("driver not found")
	}

	if !account.IsDriver {
		return errors.New("only drivers can accept a ride")
	}

	ride, rideError := a.RideRepository.GetRideById(input.RideId)

	if rideError != nil {
		return rideError
	}

	if ride == nil {
		return errors.New("ride not found")
	}

	ride.Accept(input.DriverId)
	err := a.RideRepository.UpdateRide(ride)

	if err != nil {
		return err
	}

	return nil
}

File: backend/ride-express/src/application/usecase/AcceptRide.ts
import type AccountRepository from '../repository/AccountRepository';
import type RideRepository from '../repository/RideRepository';

export default class AcceptRide {
  constructor(
    private rideRepository: RideRepository,
    private accountRepository: AccountRepository
  ) {}

  async execute(input: Input) {
    const account = await this.accountRepository.getById(input.driverId);
    if (account && !account.isDriver)
      throw new Error('Only drivers can accept a ride');
    const ride = await this.rideRepository.getById(input.rideId);
    if (!ride) throw new Error('Ride not found');
    ride.accept(input.driverId);
    await this.rideRepository.update(ride);
  }
}

type Input = {
  driverId: string;
  rideId: string;
};
